File: tests/bindings/e2e/rust/api_persistence_fixture.rs
use std::collections::BTreeMap;

use statespec_generated::api_dispatcher;
use statespec_generated::api_handler_registry::DefaultApiHandler;
use statespec_generated::backend::{Backend, BackendError, BackendResult};
use statespec_generated::descriptor_types::{ApiRequestContext, ApiResponse};
use statespec_generated::json::Json;
use statespec_generated::memory_backend::InMemoryBackend;

fn api_request(api_name: &str, method: &str, path: &str, body: Json) -> ApiRequestContext {
    ApiRequestContext {
        server_name: "EntityApi".to_string(),
        api_name: api_name.to_string(),
        method: Some(method.to_string()),
        path: Some(path.to_string()),
        body,
    }
}

fn object(values: Vec<(&str, Json)>) -> Json {
    Json::Object(
        values
            .into_iter()
            .map(|(key, value)| (key.to_string(), value))
            .collect(),
    )
}

fn require_status(response: &ApiResponse, status: i32) {
    assert_eq!(response.status_code, status);
}

fn require_string(body: &Json, field: &str, expected: &str) {
    let Json::Object(values) = body else {
        panic!("expected object body");
    };
    assert_eq!(
        values.get(field),
        Some(&Json::String(expected.to_string())),
        "unexpected string field {field}",
    );
}

fn require_array_not_empty(body: &Json, field: &str) {
    let Json::Object(values) = body else {
        panic!("expected object body");
    };
    let Some(Json::Array(items)) = values.get(field) else {
        panic!("expected array field {field}");
    };
    assert!(!items.is_empty(), "expected non-empty array field {field}");
}

fn require_array_size(body: &Json, field: &str, expected: usize) {
    let Json::Object(values) = body else {
        panic!("expected object body");
    };
    let Some(Json::Array(items)) = values.get(field) else {
        panic!("expected array field {field}");
    };
    assert_eq!(items.len(), expected, "unexpected array size for {field}");
}

fn call_api<B: Backend>(
    handler: &DefaultApiHandler<B>,
    request: ApiRequestContext,
) -> BackendResult<ApiResponse> {
    let route_name = format!("{}.{}", request.server_name, request.api_name);
    api_dispatcher::dispatch_api_route(
        &handler.backend,
        handler.business_handler.as_deref(),
        &route_name,
        &request,
    )?
    .ok_or_else(|| BackendError::Internal {
        message: format!("route not handled: {route_name}"),
    })
}

#[test]
fn generated_api_persistence_handlers_round_trip_entities() {
    let handler = DefaultApiHandler::new(InMemoryBackend::new());

    let account = call_api(
        &handler,
        api_request(
            "CreateAccount",
            "POST",
            "/v1/tenants/t1/accounts/a1",
            object(vec![("display_name", Json::String("Acme".to_string()))]),
        ),
    )
    .expect("create account failed");
    require_status(&account, 201);
    require_string(&account.body, "status", "Active");

    let project = call_api(
        &handler,
        api_request(
            "CreateProject",
            "POST",
            "/v1/tenants/t1/projects/p1",
            object(vec![
                ("account_id", Json::String("a1".to_string())),
                ("name", Json::String("Core".to_string())),
            ]),
        ),
    )
    .expect("create project failed");
    require_status(&project, 201);
    require_string(&project.body, "status", "Planned");

    let task = call_api(
        &handler,
        api_request(
            "CreateTask",
            "POST",
            "/v1/tenants/t1/tasks/t1",
            object(vec![
                ("account_id", Json::String("a1".to_string())),
                ("project_id", Json::String("p1".to_string())),
                ("title", Json::String("Ship".to_string())),
                ("priority", Json::Integer(7)),
            ]),
        ),
    )
    .expect("create task failed");
    require_status(&task, 201);
    require_string(&task.body, "status", "Open");

    let other_account = call_api(
        &handler,
        api_request(
            "CreateAccount",
            "POST",
            "/v1/tenants/t1/accounts/a2",
            object(vec![("display_name", Json::String("Other".to_string()))]),
        ),
    )
    .expect("create other account failed");
    require_status(&other_account, 201);

    let other_project = call_api(
        &handler,
        api_request(
            "CreateProject",
            "POST",
            "/v1/tenants/t1/projects/p2",
            object(vec![
                ("account_id", Json::String("a2".to_string())),
                ("name", Json::String("Other".to_string())),
            ]),
        ),
    )
    .expect("create other project failed");
    require_status(&other_project, 201);

    let other_task = call_api(
        &handler,
        api_request(
            "CreateTask",
            "POST",
            "/v1/tenants/t1/tasks/t2",
            object(vec![
                ("account_id", Json::String("a2".to_string())),
                ("project_id", Json::String("p2".to_string())),
                ("title", Json::String("Other".to_string())),
                ("priority", Json::Integer(2)),
            ]),
        ),
    )
    .expect("create other task failed");
    require_status(&other_task, 201);

    let got_account = call_api(
        &handler,
        api_request(
            "GetAccount",
            "GET",
            "/v1/tenants/t1/accounts/a1",
            Json::Object(BTreeMap::new()),
        ),
    )
    .expect("get account failed");
    require_status(&got_account, 200);
    require_string(&got_account.body, "display_name", "Acme");

    let got_project = call_api(
        &handler,
        api_request(
            "GetProject",
            "GET",
            "/v1/tenants/t1/projects/p1",
            Json::Object(BTreeMap::new()),
        ),
    )
    .expect("get project failed");
    require_status(&got_project, 200);
    require_string(&got_project.body, "name", "Core");

    let got_task = call_api(
        &handler,
        api_request(
            "GetTask",
            "GET",
            "/v1/tenants/t1/tasks/t1",
            Json::Object(BTreeMap::new()),
        ),
    )
    .expect("get task failed");
    require_status(&got_task, 200);
    require_string(&got_task.body, "title", "Ship");

    let missing_account = call_api(
        &handler,
        api_request(
            "GetAccount",
            "GET",
            "/v1/tenants/t1/accounts/missing",
            Json::Object(BTreeMap::new()),
        ),
    )
    .expect("get missing account failed");
    require_status(&missing_account, 404);

    let accounts = call_api(
        &handler,
        api_request(
            "ListAccounts",
            "GET",
            "/v1/tenants/t1/accounts",
            Json::Object(BTreeMap::new()),
        ),
    )
    .expect("list accounts failed");
    require_status(&accounts, 200);
    require_array_not_empty(&accounts.body, "accounts");
    require_array_size(&accounts.body, "accounts", 2);

    let projects = call_api(
        &handler,
        api_request(
            "ListAccountProjects",
            "GET",
            "/v1/tenants/t1/accounts/a1/projects",
            Json::Object(BTreeMap::new()),
        ),
    )
    .expect("list projects failed");
    require_status(&projects, 200);
    require_array_not_empty(&projects.body, "projects");
    require_array_size(&projects.body, "projects", 1);

    let tasks = call_api(
        &handler,
        api_request(
            "ListProjectTasks",
            "GET",
            "/v1/tenants/t1/projects/p1/tasks",
            Json::Object(BTreeMap::new()),
        ),
    )
    .expect("list tasks failed");
    require_status(&tasks, 200);
    require_array_not_empty(&tasks.body, "tasks");
    require_array_size(&tasks.body, "tasks", 1);

    let active_project = call_api(
        &handler,
        api_request(
            "UpdateProjectStatus",
            "PATCH",
            "/v1/tenants/t1/projects/p1/status",
            object(vec![("status", Json::String("Active".to_string()))]),
        ),
    )
    .expect("update project failed");
    require_status(&active_project, 200);
    require_string(&active_project.body, "status", "Active");

    let in_progress_task = call_api(
        &handler,
        api_request(
            "UpdateTaskStatus",
            "PATCH",
            "/v1/tenants/t1/tasks/t1/status",
            object(vec![("status", Json::String("InProgress".to_string()))]),
        ),
    )
    .expect("update task failed");
    require_status(&in_progress_task, 200);
    require_string(&in_progress_task.body, "status", "InProgress");

    let missing_project_status = call_api(
        &handler,
        api_request(
            "UpdateProjectStatus",
            "PATCH",
            "/v1/tenants/t1/projects/missing/status",
            object(vec![("status", Json::String("Active".to_string()))]),
        ),
    )
    .expect("update missing project failed");
    require_status(&missing_project_status, 404);

    let deleted_project = call_api(
        &handler,
        api_request(
            "DeleteProject",
            "DELETE",
            "/v1/tenants/t1/projects/p1",
            Json::Object(BTreeMap::new()),
        ),
    )
    .expect("delete project failed");
    require_status(&deleted_project, 204);

    let missing_project_delete = call_api(
        &handler,
        api_request(
            "DeleteProject",
            "DELETE",
            "/v1/tenants/t1/projects/missing",
            Json::Object(BTreeMap::new()),
        ),
    )
    .expect("delete missing project failed");
    require_status(&missing_project_delete, 404);

    let got_deleted_project = call_api(
        &handler,
        api_request(
            "GetProject",
            "GET",
            "/v1/tenants/t1/projects/p1",
            Json::Object(BTreeMap::new()),
        ),
    )
    .expect("get deleted project failed");
    require_status(&got_deleted_project, 200);
    require_string(&got_deleted_project.body, "status", "Deleted");
}

File: tests/bindings/e2e/java/ApiPersistenceFixture.java
package com.statespec.generated;

import com.statespec.backend.Json;
import com.statespec.backend.memory.InMemoryBackend;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public final class ApiPersistenceFixture
{
    private ApiPersistenceFixture() {}

    public static void main(String[] args) throws Exception
    {
        ApiHandlerRegistry.DefaultHandler handler =
            ApiHandlerRegistry.defaultHandler(new InMemoryBackend());

        ApiResponse account = callApi(
            handler, request(
                         "CreateAccount", "POST", "/v1/tenants/t1/accounts/a1",
                         Json.object(Map.of("display_name", Json.string("Acme")))
                     )
        );
        requireStatus(account, 201);
        requireString(account.body(), "status", "Active");

        ApiResponse project = callApi(
            handler,
            request(
                "CreateProject", "POST", "/v1/tenants/t1/projects/p1",
                Json.object(Map.of("account_id", Json.string("a1"), "name", Json.string("Core")))
            )
        );
        requireStatus(project, 201);
        requireString(project.body(), "status", "Planned");

        ApiResponse task = callApi(
            handler, request(
                         "CreateTask", "POST", "/v1/tenants/t1/tasks/t1",
                         Json.object(Map.of(
                             "account_id", Json.string("a1"), "project_id", Json.string("p1"),
                             "title", Json.string("Ship"), "priority", Json.integer(7)
                         ))
                     )
        );
        requireStatus(task, 201);
        requireString(task.body(), "status", "Open");

        ApiResponse otherAccount = callApi(
            handler, request(
                         "CreateAccount", "POST", "/v1/tenants/t1/accounts/a2",
                         Json.object(Map.of("display_name", Json.string("Other")))
                     )
        );
        requireStatus(otherAccount, 201);

        ApiResponse otherProject = callApi(
            handler,
            request(
                "CreateProject", "POST", "/v1/tenants/t1/projects/p2",
                Json.object(Map.of("account_id", Json.string("a2"), "name", Json.string("Other")))
            )
        );
        requireStatus(otherProject, 201);

        ApiResponse otherTask = callApi(
            handler, request(
                         "CreateTask", "POST", "/v1/tenants/t1/tasks/t2",
                         Json.object(Map.of(
                             "account_id", Json.string("a2"), "project_id", Json.string("p2"),
                             "title", Json.string("Other"), "priority", Json.integer(2)
                         ))
                     )
        );
        requireStatus(otherTask, 201);

        ApiResponse gotAccount =
            callApi(handler, request("GetAccount", "GET", "/v1/tenants/t1/accounts/a1"));
        requireStatus(gotAccount, 200);
        requireString(gotAccount.body(), "display_name", "Acme");

        ApiResponse gotProject =
            callApi(handler, request("GetProject", "GET", "/v1/tenants/t1/projects/p1"));
        requireStatus(gotProject, 200);
        requireString(gotProject.body(), "name", "Core");

        ApiResponse gotTask =
            callApi(handler, request("GetTask", "GET", "/v1/tenants/t1/tasks/t1"));
        requireStatus(gotTask, 200);
        requireString(gotTask.body(), "title", "Ship");

        ApiResponse missingAccount =
            callApi(handler, request("GetAccount", "GET", "/v1/tenants/t1/accounts/missing"));
        requireStatus(missingAccount, 404);

        ApiResponse accounts =
            callApi(handler, request("ListAccounts", "GET", "/v1/tenants/t1/accounts"));
        requireStatus(accounts, 200);
        requireArrayNotEmpty(accounts.body(), "accounts");
        requireArraySize(accounts.body(), "accounts", 2);

        ApiResponse projects = callApi(
            handler, request("ListAccountProjects", "GET", "/v1/tenants/t1/accounts/a1/projects")
        );
        requireStatus(projects, 200);
        requireArrayNotEmpty(projects.body(), "projects");
        requireArraySize(projects.body(), "projects", 1);

        ApiResponse tasks = callApi(
            handler, request("ListProjectTasks", "GET", "/v1/tenants/t1/projects/p1/tasks")
        );
        requireStatus(tasks, 200);
        requireArrayNotEmpty(tasks.body(), "tasks");
        requireArraySize(tasks.body(), "tasks", 1);

        ApiResponse activeProject = callApi(
            handler, request(
                         "UpdateProjectStatus", "PATCH", "/v1/tenants/t1/projects/p1/status",
                         Json.object(Map.of("status", Json.string("Active")))
                     )
        );
        requireStatus(activeProject, 200);
        requireString(activeProject.body(), "status", "Active");

        ApiResponse inProgressTask = callApi(
            handler, request(
                         "UpdateTaskStatus", "PATCH", "/v1/tenants/t1/tasks/t1/status",
                         Json.object(Map.of("status", Json.string("InProgress")))
                     )
        );
        requireStatus(inProgressTask, 200);
        requireString(inProgressTask.body(), "status", "InProgress");

        ApiResponse missingProjectStatus = callApi(
            handler, request(
                         "UpdateProjectStatus", "PATCH", "/v1/tenants/t1/projects/missing/status",
                         Json.object(Map.of("status", Json.string("Active")))
                     )
        );
        requireStatus(missingProjectStatus, 404);

        ApiResponse deletedProject =
            callApi(handler, request("DeleteProject", "DELETE", "/v1/tenants/t1/projects/p1"));
        requireStatus(deletedProject, 204);

        ApiResponse missingProjectDelete =
            callApi(handler, request("DeleteProject", "DELETE", "/v1/tenants/t1/projects/missing"));
        requireStatus(missingProjectDelete, 404);

        ApiResponse gotDeletedProject =
            callApi(handler, request("GetProject", "GET", "/v1/tenants/t1/projects/p1"));
        requireStatus(gotDeletedProject, 200);
        requireString(gotDeletedProject.body(), "status", "Deleted");
    }

    private static ApiRequestContext request(
        String apiName,
        String method,
        String path
    )
    {
        return request(apiName, method, path, Json.nullValue());
    }

    private static ApiRequestContext request(
        String apiName,
        String method,
        String path,
        Json body
    )
    {
        return new ApiRequestContext(
            "EntityApi", apiName, Optional.of(method), Optional.of(path), body
        );
    }

    private static ApiResponse callApi(
        ApiHandlerRegistry.DefaultHandler handler,
        ApiRequestContext context
    ) throws Exception
    {
        return ApiDispatcher
            .dispatchApiRoute(
                handler.backend(), handler.businessHandler(),
                context.serverName() + "." + context.apiName(), context
            )
            .orElseGet(() -> new ApiResponse(404, Json.object(Map.of())));
    }

    private static void requireStatus(
        ApiResponse response,
        int status
    )
    {
        if (response.statusCode() != status)
        {
            throw new IllegalStateException("unexpected status code");
        }
    }

    private static void requireString(
        Json body,
        String field,
        String expected
    )
    {
        Json value = body.find(field).orElseThrow();
        if (!(value instanceof Json.StringValue stringValue) ||
            !stringValue.value().equals(expected))
        {
            throw new IllegalStateException("unexpected string field " + field);
        }
    }

    private static void requireArrayNotEmpty(
        Json body,
        String field
    )
    {
        Json value = body.find(field).orElseThrow();
        if (!(value instanceof Json.ArrayValue arrayValue) || arrayValue.values().isEmpty())
        {
            throw new IllegalStateException("expected non-empty array field " + field);
        }
    }

    private static void requireArraySize(
        Json body,
        String field,
        int expected
    )
    {
        Json value = body.find(field).orElseThrow();
        if (!(value instanceof Json.ArrayValue arrayValue) ||
            arrayValue.values().size() != expected)
        {
            throw new IllegalStateException("unexpected array size for " + field);
        }
    }
}
